src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs

```diff
@@ -43,6 +43,10 @@ public abstract partial class PipeReader
         ///     Passing a value of 0 for <paramref name="minimumSize" /> will return a <see cref="System.Threading.Tasks.ValueTask{T}" /> that will not complete until
         ///     further data is available. You should instead call <see cref="System.IO.Pipelines.PipeReader.TryRead" /> to avoid a blocking call.
         ///     </para>
+        ///     <para>
+        ///     Subsequent calls to <see cref="System.IO.Pipelines.PipeReader.AdvanceTo(System.SequencePosition,System.SequencePosition)" /> should
+        ///     examine at least <paramref name="minimumSize" /> bytes in order to avoid an <see cref="System.InvalidOperationException" />.
+        ///     </para>
```

Comment:
@davidfowl @BrennanConroy This happens because the default `ReadAtLeastAsync` implementation could have already "examined" up to the `minimumSize` already, and we don't allow moving the examined cursor back in our default implementation. It's possible that other implementations also prevent this, but I have my doubts. It doesn't look like `StreamPipeReader` prevents this.

It seems to me that the nicer thing to do would just be to allow the `examined` cursor to reverse so long as it's still ahead of the `consumed` cursor. Am I missing something?

Reply:
So basically 
```csharp
var buffer = Pipe.ReadAsync();
Pipe.AdvanceTo(buffer.Start, buffer.End);
Pipe.AdvanceTo(buffer.Start, buffer.Start);
```
would not throw anymore?

My biggest concern is about backpressure behavior that relies on Examined, but a quick glance looks like we would be increasing the length of the pipe buffer which would make backpressure more likely (which is good, don't want to release backpressure here).

It's still a super weird case to un-examine bytes though 😄

Reply:
The trick to not interfering with backpressure would be not reducing `_unconsumedBytes` when `examined` reverses, and then not double counting the bytes that get reexamined. Not super hard, but it would require at least one additional long to tracking `_currentlyUnexaminedBytes` or whatever.

Also, `_unconsumedBytes` should be renamed to `_unexaminedBytes`. The meaning of that variable changed a long time ago.